File: exercises/1-extra.dart
void main(){
    Produto cerveja = new Produto(preco: 10.50,desc: "Cerveja",data: DateTime.now());
    Produto banana = new Produto(preco: 2.00,desc: "Banana",data: DateTime.now());
    
    Item engradado = new Item(quantidade: 2, prod: cerveja);
    Item penca = new Item(quantidade: 4, prod: banana);


    Venda venda = new Venda(DateTime.now());
    // O item deve ser adicionado a lista utilizando a função
    venda.addItem(penca);
    venda.addItem(engradado);
    print(venda);
}

class Venda{
    final DateTime data;
    List<Item> itens = [];

    double total() => itens.fold(0,(soma,elemento) => soma + elemento.total());
    

    Venda(this.data);

    void addItem(Item item){
        itens.add(item);
    }

    @override
    String toString() => "\nData: ${this.data} \n Itens: (\n${this.itens}\n) \n Preço total da venda: ${this.total()}";
}

class Item{
    final double quantidade;
    Produto prod;

    double total(){
        return quantidade * prod.preco;
    }

    Item({required double this.quantidade, required Produto this.prod});

    @override
    String toString() => "\nQuantidade: ${this.quantidade} \n (Produto: ${this.prod}) Preço total do item: ${this.total()}\n -____________________________-";
}

class Produto{
    final double preco;
    final String desc;
    final DateTime data;

    Produto({required double this.preco, required String this.desc, required DateTime this.data});

    @override
    String toString() => "Preco do produto: ${this.preco} \n Desc: ${this.desc} \n Data: ${this.data}\n";
}

File: exercises/1-extra.py
from datetime import datetime

class Produto:
    def __init__(self, preco, desc, data):
        self.preco = preco
        self.desc = desc
        self.data = data
    
    def __str__(self):
        return ("Preco: ${self.preco} \n Desc: ${self.desc} \n Data: ${this.data}")

class Item: 
    def __init__(self,quantidade, produto):
        self.quantidade = quantidade
        self.produto = produto

    def total(self):
        return self.quantidade * self.produto.preco
    
    def __str__(self):
        return "\nQuantidade: ${self.quantidade} \n Produtos: ${self.prod}"

class Venda: 
    def __init__(self, data):
        self.data = data
        self.itens = [] 
    
    def total():
            return sum(map(lambda item: item.total(), this.itens))

    def __str__(self):
        return "\nData: ${self.data} \nTotal: ${self.itens}"    


cerveja = Produto(10.50,"Cerveja",datetime.now())
banana = Produto(2.00,"Banana",datetime.now())
print(cerveja)
print(banana)

engradado = Item(2, cerveja)
print(engradado)
print("Preço total item: ${engradado.total()}")

venda = Venda(datetime.now(),[engradado])
venda.setItem(2, banana)
print(venda)
print("Preço total: ${venda.total()}")
